pkg/streamingpromql/operators/functions/range_vectors.go

```diff
@@ -815,3 +816,85 @@ func quantileOverTime(step *types.RangeVectorStepData, _ float64, args []types.S
 
 	return Quantile(q, values), true, nil, nil
 }
+
+var DoubleExponentialSmoothing = FunctionOverRangeVectorDefinition{
+	SeriesMetadataFunction:         DropSeriesName,
+	StepFunc:                       doubleExponentialSmoothing,
+	NeedsSeriesNamesForAnnotations: true,
+}
+
+// doubleExponentialSmoothing is similar to a weighted moving average, where
+// historical data has exponentially less influence on the current data. It also
+// accounts for trends in data. The smoothing factor (0 < sf < 1) affects how
+// historical data will affect the current data. A lower smoothing factor
+// increases the influence of historical data. The trend factor (0 < tf < 1)
+// affects how trends in historical data will affect the current data. A higher
+// trend factor increases the influence. of trends. Algorithm taken from
+// https://en.wikipedia.org/wiki/Exponential_smoothing .
+func doubleExponentialSmoothing(step *types.RangeVectorStepData, _ float64, args []types.ScalarData, timeRange types.QueryTimeRange, emitAnnotation types.EmitAnnotationFunc, _ *limiting.MemoryConsumptionTracker) (float64, bool, *histogram.FloatHistogram, error) {
+	smoothingFactor := args[0].Samples[timeRange.PointIndex(step.StepT)].F
+	if smoothingFactor <= 0 || smoothingFactor >= 1 {
+		return 0, false, nil, fmt.Errorf("invalid smoothing factor. Expected: 0 < sf < 1, got: %f", smoothingFactor)
+	}
+
+	trendFactor := args[1].Samples[timeRange.PointIndex(step.StepT)].F
+	if trendFactor <= 0 || trendFactor >= 1 {
+		return 0, false, nil, fmt.Errorf("invalid trend factor. Expected: 0 < tf < 1, got: %f", trendFactor)
+	}
+
+	fHead, fTail := step.Floats.UnsafePoints()
+
+	if step.Floats.Any() && step.Histograms.Any() {
+		emitAnnotation(annotations.NewHistogramIgnoredInMixedRangeInfo)
+	}
+
+	if (len(fHead) + len(fTail)) < 2 {
+		return 0, false, nil, nil
+	}
+
+	var smooth0, smooth1, estimatedTrend float64
+
+	// Initial value of smooth1 is the first sample.
+	smooth1 = fHead[0].F
+
+	// Initial estimated trend is the difference between the first and second samples.
+	// First sample will always be in the head.
+	// If we have only one point in head, the second sample is the first index in tail.
+	if len(fHead) == 1 {
+		estimatedTrend = fTail[0].F - fHead[0].F
+	} else {
+		estimatedTrend = fHead[1].F - fHead[0].F
+	}
+
+	accumulate := func(samples []promql.FPoint, head bool) {
+		var x, y float64
+
+		for i, sample := range samples {
+			// Scale the raw value against the smoothing factor.
+			x = smoothingFactor * sample.F
+
+			// Scale the last smoothed value for all samples except the first one in head.
+			if head && i != 0 || !head {
+				estimatedTrend = calcTrendValue(trendFactor, smooth0, smooth1, estimatedTrend)
+			}
+			y = (1 - smoothingFactor) * (smooth1 + estimatedTrend)
+
+			smooth0, smooth1 = smooth1, x+y
+		}
+	}
+	accumulate(fHead[1:], true)
+	accumulate(fTail, false)
```

Comment:
perhaps something like?
```suggestion
	accumulate(fHead[1:])
	accumulate(fTail[fTailPos:0])
```
where fTailPos is set to either `0` or `1` at the initial value part.

Reply:
We need flag to pass to accumulate that tell the passed slice is a head. Because we need the head flag to tell when we can compute bt start from t >= 3.

![image](https://github.com/user-attachments/assets/72551288-37fd-42cb-b475-66b40c75f036)

Somehow in prometheus actually it computes b1 as b1 = 0 as I mentioned in another comment too.

```
// this is how they call it
b = calcTrendValue(i-1, tf, s0, s1, b)

// this is the upstream function
func calcTrendValue(i int, tf, s0, s1, b float64) float64 {
	if i == 0 { // when i in the caller is on the second sample (i = 1 or b1), we just return b directly 
		return b // so b1 = b0
	}
 
        // We only calculate the trend from sample 3 (i=2) onward
	x := tf * (s1 - s0)
	y := (1 - tf) * b

	return x + y
}
```

> `accumulate(fTail[fTailPos:0])` where fTailPos is set to either 0 or 1 at the initial value part.

I don't get this part. Can you elaborate more?

Reply:
Ignore my suggestion, I was trying to suggest a simplification to avoid the head check in accumulate, but did not take into account that `y` is still recalculated.

However, my concern here is the case where `len(fHead)` is equal to `1`. In that case, I think `if head && i > 0 || !head` check is incorrect as it'll be the second point, and not the third since `head = false`. Does that make sense?

I think something is missing around the edge case of the ring buffer here. A test would be good if possible.

Reply:
@jhesketh thanks for this point I think the concern is valid. I will double check this again and the necessary test.

In the meantime do you have any suggestion how to setup engine_test in such a way ring buffer will have exactly 1 point in the head?

Reply:
My approach would be to just test `doubleExponentialSmoothing`. Set up all the parameters for the function, you should be able to define a `types.RangeVectorStepData` with `FPointRingBuffer` where you set the slice very specifically

Reply:
Thanks for this suggestion, I think uncover the bug in the original implementation by adding the suggested test!

Added the test and also the fix here: 6c09841c5cdcd101aa4a053decee93ff13c68576